File: src/services/vowifi.ts
/**
 * VoWiFi (ap2004) Service Handler
 *
 * Builds the ApplicationConfig for Wi-Fi Calling entitlements.
 */

import type { ApplicationConfig } from '../protocol/responseTypes.js';
import { EntitlementStatus, TcStatus } from '../protocol/statusCodes.js';

interface VoWiFiConfigData {
  addresses?: Array<{ addrType: string; addr: string }>;
  serviceFlowUrl?: string;
}

const DEFAULT_ADDRESSES = [
  { addrType: '1', addr: 'epdg.operator.com' },
  { addrType: '1', addr: 'pcscf.operator.com' },
];

export function buildVoWiFiConfig(
  status: number,
  provStatus: number,
  tcStatus: number,
  configData?: unknown,
): ApplicationConfig {
  const config = (configData ?? {}) as VoWiFiConfigData;

  const result: ApplicationConfig = {
    appId: 'ap2004',
    entitlementStatus: status,
    tcStatus,
    provStatus,
  };

  if (status === EntitlementStatus.ENABLED) {
    result.addrStatus = 1;
    result.addresses = config.addresses ?? DEFAULT_ADDRESSES;
  } else {
    result.addrStatus = 0;
  }

  if (tcStatus === TcStatus.REQUIRES_ACCEPTANCE && config.serviceFlowUrl) {
    result.serviceFlowUrl = config.serviceFlowUrl;
  }

  return result;
}

File: internal/services/vowifi.go
package services

import (
	"github.com/hkassaei/poc-entitlements-as-a-service-go/internal/protocol"
)

// VoWiFiConfigData holds configData for ap2004.
type VoWiFiConfigData struct {
	Addresses      []protocol.AddressConfig `json:"addresses,omitempty"`
	ServiceFlowURL string                   `json:"serviceFlowUrl,omitempty"`
}

var _defaultVoWiFiAddresses = []protocol.AddressConfig{
	{AddrType: "1", Addr: "epdg.operator.com"},
	{AddrType: "1", Addr: "pcscf.operator.com"},
}

// BuildVoWiFiConfig builds the ApplicationConfig for Wi-Fi Calling (ap2004).
func BuildVoWiFiConfig(status, provStatus, tcStatus int, configData *VoWiFiConfigData) *protocol.ApplicationConfig {
	if configData == nil {
		configData = &VoWiFiConfigData{}
	}

	result := &protocol.ApplicationConfig{
		AppID:             "ap2004",
		EntitlementStatus: status,
		TcStatus:          protocol.IntPtr(tcStatus),
		ProvStatus:        protocol.IntPtr(provStatus),
	}

	if status == protocol.EntitlementStatusEnabled {
		result.AddrStatus = protocol.IntPtr(1)
		if len(configData.Addresses) > 0 {
			result.Addresses = configData.Addresses
		} else {
			result.Addresses = _defaultVoWiFiAddresses
		}
	} else {
		result.AddrStatus = protocol.IntPtr(0)
	}

	if tcStatus == protocol.TcStatusRequiresAcceptance && configData.ServiceFlowURL != "" {
		result.ServiceFlowURL = configData.ServiceFlowURL
	}

	return result
}
